indico/modules/events/surveys/controllers/display.py

```diff
@@ -112,13 +121,30 @@ def _process(self):
                                                event=self.event_new, survey=self.survey,
                                                back_button_endpoint=back_button_endpoint)
 
-    def _save_answers(self, form):
+    def _make_form(self):
+        survey_form_class = make_survey_form(self.survey)
+        if self.submission and request.method == 'GET':
+            return survey_form_class(formdata=MultiDict(self.submission.pending_answers))
+        else:
+            return survey_form_class()
+
+    @no_autoflush
+    def _save_answers(self, form, submission):
         survey = self.survey
-        submission = SurveySubmission(survey=survey)
-        if not survey.anonymous:
-            submission.user = session.user
+        if not submission:
+            submission = SurveySubmission(survey=survey, user=session.user)
+        submission.is_anonymous = survey.anonymous
         for question in survey.questions:
             answer = SurveyAnswer(question=question, data=getattr(form, 'question_{}'.format(question.id)).data)
             submission.answers.append(answer)
         db.session.flush()
         return submission
+
+
+class RHSaveSurveyAnswers(RHSubmitSurvey):
+    def _process(self):
+        pending_answers = {k: v for k, v in request.form.iterlists() if k.startswith('question_')}
+        if not self.submission:
+            self.submission = SurveySubmission(survey=self.survey, user=session.user)
+        self.submission.pending_answers = pending_answers
+        self.submission.is_anonymous = self.survey.anonymous
```

Comment:
I wonder if `is_anonymous` is even needed anymore. An incomplete submission is never shown anyway and we could just determine whether to keep the user or not based on the anonymity setting when doing the final submission.
It's quite unlikely that the setting changes while a survey is running anyway.

cc @pferreir